yarn-project/ethereum/src/l1_tx_utils.test.ts

```diff
@@ -0,0 +1,189 @@
+import getPort from 'get-port';
+import { createPublicClient, createWalletClient, http } from 'viem';
+import { mnemonicToAccount, privateKeyToAccount } from 'viem/accounts';
+import { foundry } from 'viem/chains';
+
+import { L1TxUtils } from './l1_tx_utils.js';
+
+const MNEMONIC = 'test test test test test test test test test test test junk';
+const WEI_CONST = 1_000_000_000n;
+// Simple contract that just returns 42
+const SIMPLE_CONTRACT_BYTECODE = '0x69602a60005260206000f3600052600a6016f3';
+
+const startAnvil = async (l1BlockTime?: number) => {
+  const ethereumHostPort = await getPort();
+  const rpcUrl = `http://127.0.0.1:${ethereumHostPort}`;
+  const anvil = createAnvil({
+    port: ethereumHostPort,
+    blockTime: l1BlockTime,
+  });
+  await anvil.start();
+  return { anvil, rpcUrl };
+};
+
+describe('GasUtils', () => {
+  let gasUtils: L1TxUtils;
+  let publicClient: any;
+  let walletClient: any;
+  let anvil: Anvil;
+  let initialBaseFee: bigint;
+  const logger = createDebugLogger('l1_gas_test');
+
+  beforeAll(async () => {
+    const { anvil: anvilInstance, rpcUrl } = await startAnvil(1);
+    anvil = anvilInstance;
+    const hdAccount = mnemonicToAccount(MNEMONIC, { addressIndex: 0 });
+    const privKeyRaw = hdAccount.getHdKey().privateKey;
+    if (!privKeyRaw) {
+      throw new Error('Failed to get private key');
+    }
+    const privKey = Buffer.from(privKeyRaw).toString('hex');
+    const account = privateKeyToAccount(`0x${privKey}`);
+
+    publicClient = createPublicClient({
+      transport: http(rpcUrl),
+      chain: foundry,
+    });
+
+    walletClient = createWalletClient({
+      transport: http(rpcUrl),
+      chain: foundry,
+      account,
+    });
+
+    gasUtils = new L1TxUtils(publicClient, walletClient, logger, {
+      bufferPercentage: 20n,
+      maxGwei: 500n,
+      minGwei: 1n,
+      maxAttempts: 3,
+      checkIntervalMs: 100,
+      stallTimeMs: 1000,
+    });
+  });
+  afterAll(async () => {
+    await anvil.stop();
+  }, 5000);
+
+  it('sends and monitors a simple transaction', async () => {
+    const receipt = await gasUtils.sendAndMonitorTransaction({
+      to: '0x1234567890123456789012345678901234567890',
+      data: '0x',
+      value: 0n,
+    });
+
+    expect(receipt.status).toBe('success');
+  }, 10_000);
+
+  it('handles gas price spikes by retrying with higher gas price', async () => {
+    // Get initial base fee
+    const initialBlock = await publicClient.getBlock({ blockTag: 'latest' });
+    initialBaseFee = initialBlock.baseFeePerGas ?? 0n;
+
+    // Start a transaction
+    const sendPromise = gasUtils.sendAndMonitorTransaction({
+      to: '0x1234567890123456789012345678901234567890',
+      data: '0x',
+      value: 0n,
+    });
+
+    // Spike gas price to 3x the initial base fee
+    await publicClient.transport.request({
+      method: 'anvil_setNextBlockBaseFeePerGas',
+      params: [(initialBaseFee * 3n).toString()],
+    });
```

Comment:
Can we add another test where we spike the priority fee? Not sure if anvil supports that.